capsules/src/bmp280.rs

```diff
@@ -0,0 +1,469 @@
+            dig_t1: twobyte(i2c_raw[0], i2c_raw[1]) as u16,
+            dig_t2: twobyte(i2c_raw[2], i2c_raw[3]) as i16,
+            dig_t3: twobyte(i2c_raw[4], i2c_raw[5]) as i16,
+        }
+    }
+
+    fn temp_from_raw(&self, raw_temp: u32) -> i32 {
+        let temp = raw_temp as i32; // guaranteed to succeed because raw temp has only 20 significant bits maximum.
+        let dig_t1 = self.dig_t1 as i32; // same, 16-bits
+        let dig_t2 = self.dig_t2 as i32; // same, 16-bits
+        let dig_t3 = self.dig_t3 as i32; // same, 16-bits
+                                         // From the datasheet
+        let var1 = (((temp >> 3) - (dig_t1 << 1)) * dig_t2) >> 11;
+        let a = (temp >> 4) - dig_t1;
+        let var2 = (((a * a) >> 12) * dig_t3) >> 14;
+        let t_fine = var1 + var2;
+        ((t_fine * 5) + 128) >> 8
+    }
+}
+
+/// Internal state.
+/// Each state can lead to the next on in order of appearance.
+#[derive(Clone, Copy, PartialEq, Debug)]
+enum State {
+    Uninitialized,
+    InitId,
+    /// It's not guaranteed that the MCU reset is the same as device power-on,
+    /// so an explicit reset is necessary.
+    InitResetting,
+    InitWaitingReady,
+    InitReadingCalibration,
+
+    Idle(CalibrationData),
+
+    // States related to sample readout
+    /// One-shot mode request sent
+    Configuring(CalibrationData),
+    /// Sampling takes milliseconds, so spend most of that time sleeping.
+    WaitingForAlarm(CalibrationData),
+    /// Polling for the readout to become ready.
+    Waiting(CalibrationData),
+    /// Waiting for readout to return.
+    /// This state can also lead back to Idle.
+    Reading(CalibrationData),
+
+    /// Reset cannot be attempted.
+    /// This is because reset failed before, or because the ID is mismatched.
+    IrrecoverableError,
+    /// Irrecoverable. Currently only when init fails.
+    Error,
+}
+
+/// Complies with the reading and writing protocol used by the sensor.
+struct I2cWrapper<'a> {
+    i2c: &'a dyn i2c::I2CDevice,
+}
+
+impl<'a> I2cWrapper<'a> {
+    fn write<const COUNT: usize>(
+        &self,
+        buffer: &'static mut [u8],
+        addr: Register,
+        data: [u8; COUNT],
+    ) -> Result<(), (i2c::Error, &'static mut [u8])> {
+        buffer[0] = addr as u8;
+        for (i, d) in data.iter().enumerate() {
+            buffer[i + 1] = *d;
+        }
+        self.i2c.enable();
+        self.i2c.write(buffer, COUNT as u8 + 1)
+    }
+
+    /// Requests a read into buffer.
+    /// Parse the result using `parse_read`.
+    fn read(
+        &self,
+        buffer: &'static mut [u8],
+        addr: Register,
+        count: u8,
+    ) -> Result<(), (i2c::Error, &'static mut [u8])> {
+        buffer[0] = addr as u8;
+        self.i2c.enable();
+        self.i2c.write_read(buffer, 1, count)
+    }
+
+    fn disable(&self) {
+        self.i2c.disable()
+    }
+
+    fn parse_read(buffer: &[u8], count: u8) -> &[u8] {
+        &buffer[..(count as usize)]
+    }
```

Comment:
I'm afraid I don't fully understand the purpose of this method. Is this supposed to eventually do more than just subslicing? If not, I'd avoid this layer of indirection and just the Index operator directly.

Reply:
It was there to have a single place with an intepretation of how the buffer should be parsed. I had to change it a couple of times as I built up my understanding, and it was annoying to change it everywhere. It forms a complete, coherent application layer on top of i2c under the I2cWrapper struct.

I guess we don't expect this to change in this driver, so it could be thrown away, but I think it's valuable if anyone wants to learn from this layer as a basis for another driver, it would be useful to keep it.

Reply:
> I guess we don't expect this to change in this driver, so it could be thrown away, but I think it's valuable if anyone wants to learn from this layer as a basis for another driver, it would be useful to keep it.

Okay, I now understand the basic intention, I was just wondering why you'd want a `parsing` function just for subslicing, but it's probably a coincidence that parsing is effectively just a subslicing of the buffer in for this device and there is no more complex encapsulation / message layer. I'm fine with keeping it.